core/services/pipeline/graph_test.go

```diff
@@ -182,3 +182,28 @@ func TestGraph_HasCycles(t *testing.T) {
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "cycle detected")
 }
+
+func TestGraph_ImplicitDependencies(t *testing.T) {
+	g := pipeline.NewGraph()
+	err := g.UnmarshalText([]byte(`
+		a [type=bridge];
+		b [type=multiply times=1.23 data="$(a.a1)"];
+		c [type=xyz times=1.23 input="$(b)"];
+		d [type=xyz times=1.23 check="{\"a\": $(jobSpec.id),\"b\":$(c.p)}"];
+	`))
+
+	nodes := make(map[string]int64)
+	iter := g.Nodes()
+	for iter.Next() {
+		n := iter.Node().(interface {
+			graph.Node
+			DOTID() string
+		})
+		nodes[n.DOTID()] = n.ID()
+	}
+	require.NoError(t, err)
+	require.Equal(t, 3, g.Edges().Len())
+	require.Equal(t, true, g.HasEdgeFromTo(nodes["a"], nodes["b"]))
```

Comment:
1. You can use `require.True` here to save a few keystrokes
2. For the sake of being thorough, it might be worth also testing the negative cases (e.g., `require.False(t, g.HasEdgeFromTo(nodes["a"], nodes["c"]))` etc.)

Reply:
Thanks.
1. Done.
2. I'm already validating there are only 3 edges in total. And then verifying those 3 edges exactly.